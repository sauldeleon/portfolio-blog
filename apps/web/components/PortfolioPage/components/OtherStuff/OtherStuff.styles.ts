import { styled } from 'styled-components'

import { Link } from '@sdlgr/link'
import { Label } from '@sdlgr/typography'

export const StyledAreaPeriod = styled(Label)`
  font-size: 20px;
  margin-bottom: 20px;
  ${({ theme }) => theme.fontStyles.robotoMono.regular};
`

export const StyledList = styled.ul`
  margin-bottom: 50px;
`

export const StyledListItem = styled.li`
  font-size: 16px;
  text-indent: -20px;
  padding-left: 20px;

  &:before {
    content: 'Â·';
    font-size: 30px;
    vertical-align: middle;
    line-height: 20px;
  }
`

export const StyledItalic = styled.i`
  ${({ theme }) => theme.typography.body.M}
`

export const StyledLink = styled(Link)`
  font-weight: bold;
`
