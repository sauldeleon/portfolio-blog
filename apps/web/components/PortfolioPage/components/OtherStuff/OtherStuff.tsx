import { ParseKeys } from 'i18next'
import { Trans } from 'react-i18next'

import { useClientTranslation } from '@web/i18n/client'

import {
  StyledAreaPeriod,
  StyledItalic,
  StyledLink,
  StyledList,
  StyledListItem,
} from './OtherStuff.styles'

export function OtherStuff() {
  const { t } = useClientTranslation('portfolioPage')
  const otherItems = t('items.other.otherAreas', { returnObjects: true })

  return (
    <>
      {otherItems.map(({ name, beginYear, endYear, highlights }) => (
        <div key={name}>
          <StyledAreaPeriod>
            <Trans
              t={t}
              i18nKey={name as ParseKeys<'portfolioPage'>}
              components={{
                italic: <StyledItalic />,
              }}
            />{' '}
            {`${beginYear || ''}${endYear ? ' - ' + endYear : ''}`}
          </StyledAreaPeriod>
          <StyledList>
            {highlights.map((highlight) => (
              <StyledListItem key={highlight}>
                <Trans
                  t={t}
                  i18nKey={highlight as ParseKeys<'portfolioPage'>}
                  components={{
                    linkComponent: (
                      <StyledLink
                        href="https://rd.springer.com/chapter/10.1007/978-3-642-21350-2_29#page-1"
                        target="_blank"
                      >
                        text
                      </StyledLink>
                    ),
                  }}
                />
              </StyledListItem>
            ))}
          </StyledList>
        </div>
      ))}
    </>
  )
}
