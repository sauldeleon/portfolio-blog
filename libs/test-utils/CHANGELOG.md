# Changelog Test utils

## [0.2.3](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.2...test-utils-0.2.3) (2024-01-08)

## [0.2.2](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.1...test-utils-0.2.2) (2023-09-26)

## [0.2.1](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.0...test-utils-0.2.1) (2023-09-14)

# [0.2.0](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.1.0...test-utils-0.2.0) (2023-09-05)

### Bug Fixes

- reduced duplicated lines ([f51168a](https://github.com/sauldeleon/portfolio-blog/commit/f51168aefe025a5a65ffdb2dc22f7a16b863406e))
- rename commit message format ([e1e2ab4](https://github.com/sauldeleon/portfolio-blog/commit/e1e2ab404bbd2c32f3508d1ed8197b3fbff93cb9))

### Features

- deleted changelogs ([e9be338](https://github.com/sauldeleon/portfolio-blog/commit/e9be33836ee47b6505ad94d21f4be21855a7fa0d))

# 0.1.0 (2023-09-02)

### Bug Fixes

- rename commit message format ([e1e2ab4](https://github.com/sauldeleon/portfolio-blog/commit/e1e2ab404bbd2c32f3508d1ed8197b3fbff93cb9))

### Features

- add contact initial page ([fea862b](https://github.com/sauldeleon/portfolio-blog/commit/fea862b3507268013149050ae3f05fa517e7b3d9))
- add homepage ([60523aa](https://github.com/sauldeleon/portfolio-blog/commit/60523aa31ec088d318745ef0c757b4cbe656f0d4))
- deleted changelogs ([e9be338](https://github.com/sauldeleon/portfolio-blog/commit/e9be33836ee47b6505ad94d21f4be21855a7fa0d))
- improved a11y and styles ([873f656](https://github.com/sauldeleon/portfolio-blog/commit/873f656587f50df8e29ad526fdc68fea0b2c9784))
- toggle language button ([afd97e9](https://github.com/sauldeleon/portfolio-blog/commit/afd97e95beac0f70ed643c753e71db2ca014054b))
