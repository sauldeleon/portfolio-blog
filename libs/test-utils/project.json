{
  "name": "test-utils",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/test-utils/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "version": {
      "executor": "@jscutlery/semver:version",
      "options": {
        "changelogHeader": "# Changelog Test utils \n",
        "baseBranch": "main",
        "commitMessageFormat": "chore(${projectName}): release: v${version}",
        "push": false
      }
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/test-utils/**/*.{ts,tsx,js,jsx}"]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/test-utils/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  }
}
