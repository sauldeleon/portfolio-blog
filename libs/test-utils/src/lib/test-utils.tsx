import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { Matcher } from '@testing-library/dom/types/matches'
import {
  RenderOptions,
  RenderResult,
  queryAllByAttribute as _queryAllByAttribute,
  queryByAttribute as _queryByAttribute,
  act,
  render,
  screen,
} from '@testing-library/react'
import { RouterContext } from 'next/dist/shared/lib/router-context'
import Router, { NextRouter } from 'next/router'
import { ReactNode, useEffect, useState } from 'react'
import { ThemeProvider } from 'styled-components'

import { LanguageContextProvider } from '@sdlgr/i18n-config'
import { mainTheme } from '@sdlgr/main-theme'

export type MatcherCustomFnParams = {
  content: string
  element: Element | null
}

/* istanbul ignore next */
export const renderWithTheme = (children: ReactNode): RenderResult => {
  return render(<ThemeProvider theme={mainTheme}>{children}</ThemeProvider>)
}

/* istanbul ignore next */
export function RenderProviders({ children }: { children: ReactNode }) {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

  return (
    <QueryClientProvider client={queryClient}>
      <LanguageContextProvider value={{ language: 'en' }}>
        <ThemeProvider theme={mainTheme}>{children}</ThemeProvider>
      </LanguageContextProvider>
    </QueryClientProvider>
  )
}

/* istanbul ignore next */
export const renderApp = (
  children: ReactNode,
  options?: Omit<RenderOptions, 'queries'>
) => {
  return render(<RenderProviders>{children}</RenderProviders>, options)
}

/**
 * Generic function to get element by attribute and value
 * @example queryByAttribute('id')(baseElement, 'element-id')
 * @param {string} attr
 * @return {any}
 */
export function queryByAttribute(attr: string) {
  return _queryByAttribute.bind(null, attr)
}

/**
 * Generic function to get all elements by attribute and value
 * @example queryByAttribute('class')(baseElement, 'element-class')
 * @param {string} attr
 * @returns {any}
 */
export function queryAllByAttribute(attr: string) {
  return _queryAllByAttribute.bind(null, attr)
}

/**
 * Generic function to get element by attribute and value
 * @example getByAttribute('id')(baseElement, 'element-id')
 * @param {string} attr
 * @return {any}
 */
export function getByAttribute(attr: string) {
  return _queryByAttribute.bind(null, attr)
}

/**
 * Generic function to get all elements by attribute and value
 * @example getByAttribute('class')(baseElement, 'element-class')
 * @param {string} attr
 * @returns {any}
 */
export function getAllByAttribute(attr: string) {
  return _queryAllByAttribute.bind(null, attr)
}

/**
 * Shortcut version for queryByAttribute, but specific for id
 * @param {HTMLElement} container
 * @param {string} id
 * @return {any}
 */
export function getById(container: HTMLElement, id: string) {
  return queryByAttribute('id')(container, id) as Element
}

/**
 * Shortcut version for queryByAttribute, but specific for class name
 * @param {HTMLElement} container
 * @param {string} className
 * @return {any}
 */
export function getByClassName(container: HTMLElement, className: string) {
  return queryByAttribute('class')(container, className) as Element
}

function matcherCustomFn(
  tag: string,
  matcher: Matcher,
  { element, content }: MatcherCustomFnParams
) {
  const isTag = element?.tagName.toUpperCase() === tag.toUpperCase()
  let matching = false
  if (isTag) {
    matching = content === matcher
    /* istanbul ignore else*/
    if (typeof matcher === 'function') {
      matching = matcher(content, element)
    } else if (matcher instanceof RegExp) {
      matching = matcher.test(content)
    }
  }
  return matching && isTag
}

/**
 * Get 1 element by tag and text
 * @param {string} tag
 * @param {Matcher} matcher
 * @return {any}
 */
export function getByTagAndText(tag: string, matcher: Matcher) {
  return screen.getByText((content, element) =>
    matcherCustomFn(tag, matcher, { content, element })
  )
}

/**
 * Get multiple elements by tag and text
 * @param {string} tag
 * @param {Matcher} matcher
 * @return {any}
 */
export function getAllByTagAndText(tag: string, matcher: Matcher) {
  return screen.getAllByText((content, element) =>
    matcherCustomFn(tag, matcher, { content, element })
  )
}

/**
 * Mock Math.random() to return a specific value
 * @param {number} value
 */
export function mockRandom(value = 0.1234) {
  const mockMath = Object.create(global.Math)
  mockMath.random = () => value
  global.Math = mockMath
}

/**
 * Hook to use in test to ensure "last" render is called
 */
export function useWithRefreshInTest() {
  const [withRefreshInTest, setWithRefreshInTest] = useState(false)

  useEffect(() => {
    // Change state to "rerender" the component
    setWithRefreshInTest(true)
  }, [])

  return withRefreshInTest
}

export { render } from '@testing-library/react'

/* istanbul ignore next */
export const awaitAllAsync = async () =>
  act(() => new Promise((resolve) => setTimeout(resolve, 0)))

/* istanbul ignore next */
export function renderWithNextRouter(
  children: ReactNode,
  options?: Partial<NextRouter>
) {
  return renderWithTheme(
    <RouterContext.Provider value={{ ...Router, ...options }}>
      {children}
    </RouterContext.Provider>
  )
}
