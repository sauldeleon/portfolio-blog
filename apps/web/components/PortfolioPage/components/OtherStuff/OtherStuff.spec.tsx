import { screen } from '@testing-library/react'

import { renderApp } from '@sdlgr/test-utils'

import { OtherStuff } from './OtherStuff'

describe('OtherStuff', () => {
  it('should render successfully', async () => {
    const { baseElement } = renderApp(<OtherStuff />)
    expect(await screen.findByText('Bachelor thesis')).toBeInTheDocument()

    expect(baseElement).toMatchSnapshot()
  })
})
