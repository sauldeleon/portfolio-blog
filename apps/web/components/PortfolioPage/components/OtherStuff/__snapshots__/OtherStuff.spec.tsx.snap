// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OtherStuff should render successfully 1`] = `
.c4 {
  position: relative;
  position: relative;
}

.c4::after {
  content: '';
  position: absolute;
  width: 100%;
  height: 2px;
  bottom: 0;
  left: 0;
  background-color: currentColor;
}

.c4::after {
  transform: scaleX(0);
  transform-origin: bottom right;
  transition: transform 0.25s ease-out;
}

.c4:hover::after {
  transform: scaleX(1);
  transform-origin: bottom left;
}

.c4:hover,
.c4:active,
.c4:focus {
  text-decoration: none;
}

.c4:focus-visible {
  outline: #FBFBFB auto 1px;
}

.c0 {
  font-family: fontFamily;
  font-weight: 500;
  color: #FBFBFB;
  font-size: 1rem;
  line-height: normal;
  text-transform: uppercase;
}

.c1 {
  font-size: 20px;
  margin-bottom: 20px;
  font-family: fontFamily;
  font-weight: 400;
}

.c2 {
  margin-bottom: 50px;
}

.c3 {
  font-size: 16px;
  text-indent: -20px;
  padding-left: 20px;
}

.c3:before {
  content: '·';
  font-size: 30px;
  vertical-align: middle;
  line-height: 20px;
}

.c6 {
  font-family: fontFamily;
  font-weight: 400;
  color: #FBFBFB;
  font-size: 0.9rem;
  line-height: normal;
}

.c5 {
  font-weight: bold;
}

<body>
  <div>
    <div>
      <div
        class="c0 c1"
      >
        Bachelor thesis
         
      </div>
      <ul
        class="c2"
      >
        <li
          class="c3"
        >
          Development of a support tool for students related to Data Structures for the verification and derivation of iterative algorithms using semantic tableaux with a graphical interface in Java.
        </li>
        <li
          class="c3"
        >
          This work resulted in two 
          <a
            class="c4 c5"
            data-testid="plain-anchor"
            href="https://rd.springer.com/chapter/10.1007/978-3-642-21350-2_29#page-1"
            rel="noreferrer"
            target="_blank"
          >
            papers
          </a>
          .
        </li>
      </ul>
    </div>
    <div>
      <div
        class="c0 c1"
      >
        Collaborative activities with the 
        <i
          class="c6"
        >
          Departamento de Sistemas Informáticos y Computación
        </i>
         (DSIC) at the Facultad de Informática at the Universidad Complutense of Madrid.
         
        2011 - 2012
      </div>
      <ul
        class="c2"
      >
        <li
          class="c3"
        >
          Maintenance of the Virtual Area by updating the database with information related to classes and publications.
        </li>
        <li
          class="c3"
        >
          Development of an application for counting students attending classes for obtaining statistics.
        </li>
      </ul>
    </div>
    <div>
      <div
        class="c0 c1"
      >
        This Portfolio.
         
        2022 - Until I throw it out the window
      </div>
      <ul
        class="c2"
      >
        <li
          class="c3"
        >
          Check the 'About' link in the footer of this website.
        </li>
      </ul>
    </div>
    <div>
      <div
        class="c0 c1"
      >
        MATHS Academy Teacher.
         
        2009 - 2013
      </div>
      <ul
        class="c2"
      >
        <li
          class="c3"
        >
          Taught 1st and 2nd-year Computer Engineering classes in Java and OOP.
        </li>
      </ul>
    </div>
    <div>
      <div
        class="c0 c1"
      >
        Tutor.
         
        2004 - 2016
      </div>
      <ul
        class="c2"
      >
        <li
          class="c3"
        >
          Private tutor for school and high school students in mathematics, physics, and chemistry. Strong teaching abilities.
        </li>
      </ul>
    </div>
  </div>
</body>
`;
